benchmarks/Jumps/Spatial_Signaling_Sanft.jmd

```diff
@@ -0,0 +1,240 @@
+Let's look at one example to check our model seems reasonable. We'll plot the
+total number of molecules in the system to verify we get around 28,000
+molecules, as reported in Sanft [^1], when using a domain length of 6 μm.
+```julia
+end_time = 3.0
+grid, hopping_constants, h, u0 = transport_model(rn, 60)
+majumps, dprobwm, jprobwm, u0wm = wellmixed_model(rn, u0, end_time, 6.0)
+sol = solve(jprobwm, SSAStepper(); saveat = end_time/200)
+Ntot = [sum(u) for u in sol.u]
+plt = plot(sol.t, Ntot, label="Well-mixed", ylabel="Total Number of Molecules",
+                        xlabel="time")
+
+# spatial model
+majumps, dprobwm, jprobwm, u0wm = wellmixed_model(rn, u0, end_time, h)
+dprob = DiscreteProblem(u0, (0.0, end_time), copy(dprobwm.p))
+jprob = JumpProblem(dprob, DirectCRDirect(), majumps; hopping_constants,
+                    spatial_system = grid, save_positions = (false, false))
+spatial_sol = solve(jprob, SSAStepper(); saveat = end_time/200)
+Ntot = [sum(vec(u)) for u in spatial_sol.u]
+plot!(plt, spatial_sol.t, Ntot, label="Spatial",
+      title="Steady-state number of molecules is $(Ntot[end])")
+```
+
+# Benchmarking performance of the methods
+We can now run the solvers and record the performance with `BenchmarkTools`.
+Let's first create a `DiscreteCallback` to terminate simulations once we reach
+`10^8` events:
+```julia
+@Base.kwdef mutable struct EventCallback
+    n::Int = 0
+end
+
+function (ecb::EventCallback)(u, t, integ)
+    ecb.n += 1
+    ecb.n == 10^8
+end
+
+function (ecb::EventCallback)(integ)
+    # save the final state
+    terminate!(integ)
+    nothing
+end
+```
+We next create a function to run and return our benchmarking results.
+```julia
+function benchmark_and_save!(bench_dict, end_times, Nv, algs, domain_len)
+    @assert length(end_times) == length(Nv)
+
+    # callback for terminating simulations
+    ecb = EventCallback()
+    cb = DiscreteCallback(ecb, ecb)
+
+    for (end_time, N) in zip(end_times, Nv)
+        names = ["$s"[1:end-2] for s in algs]
+
+        grid, hopping_constants, h, u0 = transport_model(rn, N; domain_len)
+
+        # we create a well-mixed model within a domain of the size of *one* voxel, h
+        majumps, dprobwm, jprobwm, u0wm = wellmixed_model(rn, u0, end_time, h)
+
+        # the spatial problem
+        dprob = DiscreteProblem(u0, (0.0, end_time), copy(dprobwm.p))
+
+        @show N
+
+        # benchmarking and saving
+        benchmarks = Vector{BenchmarkTools.Trial}(undef, length(algs))
+
+        # callback for terminating simulations
+
+        for (i, alg) in enumerate(algs)
+            name = names[i]
+            println("benchmarking $name")
+            jp = JumpProblem(dprob, alg, majumps, hopping_constants=hopping_constants,
+                             spatial_system = grid, save_positions=(false,false))
+            b = @benchmarkable solve($jp, SSAStepper(); saveat = $(dprob.tspan[2]), callback) setup = (callback = deepcopy($cb)) samples = 10 seconds = 1500
```

Comment:
is 1500 seconds enough to reach 10^8 simulation steps?

Reply:
When I went up to a mesh of N=120 I only needed ~60 seconds for the median, so ~600 seconds for 10. So I think it should be enough. That said, I'll bump it to an hour.

Reply:
And of course that is on my laptop; I have no idea how it will perform on the server.